src/commonMain/kotlin/fr/acinq/lightning/crypto/KeyManager.kt

```diff
@@ -1,128 +1,103 @@
-     *
-     * @param tx           input transaction
-     * @param privateKey   private key
-     * @param remoteSecret remote secret
-     * @return a signature generated with a private key generated from the input private key and the remote secret.
+     * Secrets and keys for a given channel.
+     * How these keys are generated depends on the [KeyManager] implementation.
      */
-    fun sign(tx: TransactionWithInputInfo, privateKey: PrivateKey, remoteSecret: PrivateKey): ByteVector64
-
-    companion object {
-        /**
-         * Create a BIP32 path from a public key. This path will be used to derive channel keys.
-         * Having channel keys derived from the funding public keys makes it very easy to retrieve your funds when've you've lost your data:
-         * - connect to your peer and use DLP to get them to publish their remote commit tx
-         * - retrieve the commit tx from the bitcoin network, extract your funding pubkey from its witness data
-         * - recompute your channel keys and spend your output
-         *
-         * @param fundingPubKey funding public key
-         * @return a BIP32 path
-         */
-        fun channelKeyPath(fundingPubKey: PublicKey): KeyPath {
-            val buffer = sha256(fundingPubKey.value)
-
-            val path = sequence {
-                var i = 0
-                while (true) {
-                    yield(Pack.int32BE(buffer, i).toUInt().toLong())
-                    i += 4
-                }
-            }
-            return KeyPath(path.take(8).toList())
+    data class ChannelKeys(
+        val fundingKeyPath: KeyPath,
+        val fundingPrivateKey: PrivateKey,
+        val paymentKey: PrivateKey,
+        val delayedPaymentKey: PrivateKey,
+        val htlcKey: PrivateKey,
+        val revocationKey: PrivateKey,
+        val shaSeed: ByteVector32
+    ) {
+        val fundingPubKey: PublicKey = fundingPrivateKey.publicKey()
+        val htlcBasepoint: PublicKey = htlcKey.publicKey()
+        val paymentBasepoint: PublicKey = paymentKey.publicKey()
+        val delayedPaymentBasepoint: PublicKey = delayedPaymentKey.publicKey()
+        val revocationBasepoint: PublicKey = revocationKey.publicKey()
+        val temporaryChannelId: ByteVector32 = (ByteVector(ByteArray(33) { 0 }) + revocationBasepoint.value).sha256()
+        fun commitmentPoint(index: Long): PublicKey = Bolt3Derivation.perCommitPoint(shaSeed, index)
+        fun commitmentSecret(index: Long): PrivateKey = Bolt3Derivation.perCommitSecret(shaSeed, index)
+    }
+
+    data class Bip84OnChainKeys(
+        private val chain: NodeParams.Chain,
+        val account: Long,
+        val xpriv: DeterministicWallet.ExtendedPrivateKey
+    ) {
+        constructor(chain: NodeParams.Chain, master: DeterministicWallet.ExtendedPrivateKey, account: Long) : this(
+            chain, account,
+            xpriv = DeterministicWallet.derivePrivateKey(master, bip84BasePath(chain) I hardened(account))
+        )
```

Comment:
I think there's a confusion between this constructor and the default one, since they take the same parameters but in a different order and this one actually does one hardened derivation using `account`. For example, `val xpub` re-derives using `account`, which means we derived twice, which is incorrect?

Do we really need to keep the `account` field in the `data class`? Can't we have the following instead?

```kotlin
data class Bip84OnChainKeys(
        private val chain: NodeParams.Chain,
        val xpriv: DeterministicWallet.ExtendedPrivateKey
    ) {
        companion object {
            fun derive(chain: NodeParams.Chain, master: DeterministicWallet.ExtendedPrivateKey, account: Long) = Bip84OnChainKeys(
                chain,
                DeterministicWallet.derivePrivateKey(master, bip84BasePath(chain) I hardened(account))    
            )
        }
    }
```

Reply:
I think it's useful to expose the account to the library integrators. I refactored in 5dfb0556, which cleared up the confusion.